### ISSUE TITLE
SMW requirement should be removed from all SMW-based extensions' composer.json

### ISSUE THREAD

**Issue by yaronkoren** [SemanticMediaWiki/SemanticMediaWiki#5892]
This is an omnibus issue, so we don't have to have the same discussion in many places. There has been a tendency to add Semantic MediaWiki to the "require" section of composer.json for each SMW-based extension. On the surface this makes sense: SMW is a requirement, after all - and this line also allows you to specify which SMW version(s) each extension is compatible with, so Composer can let you know right away if there is an incompatibility.

The problem comes if you try to download each extension via Git instead of Composer, then call "composer update" on each one to get its needed dependencies. If you try to do this with Semantic MediaWiki and Semantic Result Formats, then add both to LocalSettings.php, you will get the following error on your wiki:

> Fatal error: Cannot redeclare smwfContLang() (previously declared in .../w/extensions/SemanticMediaWiki/includes/GlobalFunctions.php:27) in .../w/extensions/SemanticResultFormats/extensions/SemanticMediaWiki/includes/GlobalFunctions.php on line 27

That's because SMW is now being included twice.

So what about just removing the require_once() call for Semantic MediaWiki? That seems like it should work, but for some reason it leads to this other error message:

> Fatal error: Uncaught MediaWiki\Registration\ExtensionDependencyError: SemanticResultFormats requires SemanticMediaWiki to be installed. in .../w/includes/registration/ExtensionRegistry.php:478

This is a problem for the Canasta package, which downloads each extension via Git, then calls "composer update" on them as necessary. Right now it has to patch each SMW-based extension it includes, to remove this line. And now that MW 1.43 has been released and a new Canasta version is in the works, we are running into this problem all over again.

Basically, I think adding SMW as a Composer requirement is unnecessary and perhaps "too clever". No one thinks "let me install Semantic Interlanguage Links, which seems like a useful extension... hm, I wonder if it requires any other extensions". SMW always comes first, and there's no way any admin will forget to download it, if they plan to use it.

I created a [pull request](https://github.com/SemanticMediaWiki/SemanticResultFormats/pull/787) about a year ago, to remove the SMW requirement from Semantic Result Formats' composer.json. There was discussion about the patch, but no firm conclusion, and in the meantime that patch has gone slightly out-of-date.

Here is a possibly non-comprehensive list of the SMW-based extensions marked as "stable" where composer.json includes a requirement for SMW:
- https://github.com/SemanticMediaWiki/SemanticCite/blob/master/composer.json#L31
- https://github.com/SemanticMediaWiki/SemanticCompoundQueries/blob/master/composer.json#L30
- https://github.com/SemanticMediaWiki/SemanticExtraSpecialProperties/blob/master/composer.json#L48
- https://github.com/SemanticMediaWiki/SemanticFormsSelect/blob/master/composer.json#L42 (also includes a line for Page Forms, which should also be removed)
- https://github.com/SemanticMediaWiki/SemanticGlossary/blob/master/composer.json#L36
- https://github.com/SemanticMediaWiki/SemanticImageCaption/blob/master/composer.json#L30
- https://github.com/SemanticMediaWiki/SemanticInterlanguageLinks/blob/master/composer.json#L31
- https://github.com/SemanticMediaWiki/SemanticMetaTags/blob/master/composer.json#L32
- https://github.com/SemanticMediaWiki/SemanticResultFormats/blob/master/composer.json#L49
- https://github.com/SemanticMediaWiki/SemanticScribunto/blob/master/composer.json#L36
- https://github.com/SemanticMediaWiki/SemanticWatchlist/blob/master/composer.json#L30

On the other hand, here are the stable SMW-based extensions I found which have no such requirements:
- https://github.com/wikimedia/mediawiki-extensions-CognitiveProcessDesigner/blob/master/composer.json
- https://github.com/ProfessionalWiki/ModernTimeline/blob/master/composer.json
- https://github.com/gesinn-it-pub/SemanticDependencyUpdater/blob/master/composer.json
- https://github.com/SemanticMediaWiki/SemanticDrilldown/blob/master/composer.json
- https://github.com/wikimedia/mediawiki-extensions-SemanticRESTAPI/blob/master/composer.json

I think the fact that these extensions all seem to have been working fine is basically proof that this requirement is not necessary.

I'm happy to create PRs for each of these 11 extensions, to remove this line (or in the case of Semantic Forms Select, two lines), if there's consensus that these lines can/should indeed be removed.

Any thoughts?

**Comment by alistair3149** [comment]
I agree with the notion to drop the SMW requirements. It'll make SMW and SMW extensions easier to install. The current implementation forces wikis to use composer to set up the SMW extensions, which unnecessarily complicates the setup process as extensions are more commonly set up through other methods (e.g. Git, drag and drop).

**Comment by JeroenDeDauw** [comment]
@yaronkoren how likely do you think it is that I will agree with you on this?

Place your bets!

Turns out I do agree about dropping the requirement from the composer.json files.

I recently observed how various SMW extensions had issues due to the SMW 5.x bump, despite not being affected by any code changes, and how that "issue" did not occur for ModernTimeline, as this SMW extension does not have the requirement in composer.json.

**Comment by yaronkoren** [comment]
I'm not sure how I would have bet on that one. :) But I'm glad you agree!

That's true, and it's a weakness of the SMW requirements in composer.json that I didn't even mention: the actual versions specified are all over the place, so a new version (like 5.0) will probably fail with about half of them, completely randomly.

**Comment by gesinn-it-gea** [comment]
I also agree with Yaron's point and have had more problems than benefits in the past by setting the dependency in composer.json.

**Comment by gesinn-it-gea** [comment]
@paladox, @gesinn-it-ilm fyi, I think you stumbled across this issue in various situations recently while fixing CI and tests...

**Comment by yaronkoren** [comment]
Okay, great - it sounds like there's consensus. I just restored the SRF patch - I'm hoping that it can be merged in, and that it will be the first of 11 or so.

**Comment by yaronkoren** [comment]
Well, I think I've done all I can easily do with this. Of the original 11 extensions I listed, six have now had their SMW requirement (and in one case, the Page Forms requirement also) removed from their composer.json files - thanks to @alistair3149 for all the merges. I created a patch for four more of them, but these are all failing validation, due to some CI setup issue. And the remaining extension is Semantic Glossary - I wasn't sure what to do with it, because its composer.json file includes requirements for both SMW and the Lingo extension. Anyway, this is good enough for me - from the Canasta perspective, I think the extensions that needed addressing have all been addressed.

**Comment by alex-mashin** [comment]
My image does `rm -rf "extensions/$DIR/extensions" && ln -s "$PWD/extensions" "extensions/$DIR/extensions"`.

**Comment by alex-mashin** [comment]
What if Composer can be somehow instructed where to install some dependecies (i.e., `../extensions` rather than `./extensions`)?

**Comment by yaronkoren** [comment]
> What if Composer can be somehow instructed where to install some dependecies (i.e., `../extensions` rather than `./extensions`)?

I hope that is not possible! Being able to modify directories other than the current one would be useful for this one purpose, but could cause all sorts of destruction in general, I would think.

**Comment by yaronkoren** [comment]
This is great! Thank you to everyone for their help. I should note that there is still one extension that includes SMW as a requirement in its composer.json file: Semantic Glossary. I didn't create a patch for that one because SG also includes the Lingo extension in composer.json, and I wasn't sure whether both should be removed or not. (Probably.) Anyway, just one left, so maybe not a big deal.

**Comment by freephile** [comment]
It looks like this effort should be expanded to also remove the similar requirement from all SMW-based extensions' composer.json in the **`require-dev`** section.

There are at least [5 repos affected](https://github.com/search?q=org%3ASemanticMediaWiki+%22mediawiki%2Fsemantic-media-wiki%22%3A+%22%40dev%22%2C&type=code) (searching for `"mediawiki/semantic-media-wiki": "@dev"`). A few repos listed above as 'fixed' are affected.

- SemanticCompoundQueries
- SemanticFormsSelect
- SemanticNotifications
- SemanticResultFormats
- SemanticScribunto

**Comment by freephile** [comment]
Also, it looks like this (original) issue affects the following repos
- [SemanticApprovedRevs](https://github.com/SemanticMediaWiki/SemanticApprovedRevs/blob/941899abf2e450ac5be0732024619aec5238f914/composer.json#L30)
- [SemanticBreadcrumbLinks](https://github.com/SemanticMediaWiki/SemanticBreadcrumbLinks/blob/fb87b0422bc1181f4994010e6f5315a0c0512cec/composer.json#L31)
- [SemanticExternalQueryLookup](https://github.com/SemanticMediaWiki/SemanticExternalQueryLookup/blob/2d041949a7ff5ddf9762d0b02652a1d14f0e914e/composer.json#L30)
- [SemanticSignup](https://github.com/SemanticMediaWiki/SemanticSignup/blob/8623231d37f9c12e006d470b17d15a115afa66ae/composer.json#L28) - an ancient extension
- [SemanticTasks](https://github.com/SemanticMediaWiki/SemanticTasks/blob/3ffd26beb69f3a573f844ccaa793d2ab7f7a70d4/composer.json#L13)

Plus advice in the SMW [docs/architecture/developing.extension.md](https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/ffba0d82cf5884b5db57b1f069d949205bcbf189/docs/architecture/developing.extension.md?plain=1#L56)
